aiida_quantumespresso/common/workchain/base/restart.py

```diff
@@ -310,13 +267,24 @@ def _handle_calculation_failure(self, calculation):
 
         return
 
-    def _prepare_process_inputs(self, process, inputs):
-        """
-        Prepare the inputs dictionary for a calculation process. Any remaining bare dictionaries in the inputs
-        dictionary will be wrapped in a Dict data node except for the 'options' key which should remain
-        a standard dictionary. Another exception are dictionaries whose keys are not strings but for example tuples.
-        This is the format used by input groups as in for example the explicit pseudo dictionary where the key is
-        a tuple of kind to which the UpfData corresponds.
+    def _handle_unexpected_failure(self, calculation, exception=None):
+        """Handle an unexpected failure.
+
+        This occurs when a calculation excepted, was killed or finished with a non-zero exit status but no errors were
+        handled. If this is the second consecutive unexpected failure the work chain is aborted.
+
+        :param calculation: the calculation that failed in an unexpected way
+        :param exception: optional exception message to log to the report
+        :return: `ExitCode` if this is the second consecutive unexpected failure
         """
-        from aiida_quantumespresso.utils.mapping import prepare_process_inputs
-        return prepare_process_inputs(process, inputs)
+        if exception:
+            self.report('{}'.format(exception))
+
+        if self.ctx.unexpected_failure:
```

Comment:
I wonder: this flag is never reset, so technically we return `ERROR_SECOND_CONSECUTIVE_UNHANDLED_FAILURE` even if this happens twice non-consecutively. Can this really happen?

Reply:
Nice catch! This is indeed a bug. I have added the fix by resetting the flag in the handle failure method as soon as a handler returns that it handled an error

Reply:
Nice, but shouldn't we also reset it if `calculation.is_finished_ok`? I would add it as first instruction here: https://github.com/aiidateam/aiida-quantumespresso/pull/325/files#diff-a075396b80333fe72d56cb4808da80caR137

Reply:
Yes, you are right, but we have to keep the other one as well. If we first have an unexpected failure, but then the second calculation has a failure but *it is* handled, then that should also reset the "unexpected failure" counter.

Originally, this was the only place where it should be reset, but since I changed the `is_finished_ok` branch to include the sanity checks, that can fail, which will cause another calculation to be run (whereas normally the work chain would terminate). In that case you are right, we should reset the flag, thanks for pointing that out! I will make the change

Reply:
I see, yes. Thank you.